aws/product/awsec2product.go

```diff
@@ -0,0 +1,179 @@
+	dbAwsProduct.InstanceType = attributes.InstanceType
+	if attributes.CurrentGeneration == "Yes" {
+		dbAwsProduct.CurrentGeneration = 1
+	}
+	if val, err := strconv.Atoi(attributes.Vcpu); err == nil {
+		dbAwsProduct.Vcpu = val
+	}
+	dbAwsProduct.Memory = attributes.Memory
+	dbAwsProduct.Storage = attributes.Storage
+	dbAwsProduct.NetworkPerformance = attributes.NetworkPerformance
+	dbAwsProduct.Tenancy = attributes.Tenancy
+	dbAwsProduct.OperatingSystem = attributes.OperatingSystem
+	if val, err := strconv.Atoi(attributes.Ecu); err == nil {
+		dbAwsProduct.Ecu = val
+	}
+	if attributes.LocationType == "AWS Region" {
+		err = storeRegionID(attributes.Location, dbAwsProduct, logger, sqlTx)
+	}
+	return err
+}
+
+// importResult parses the body returned by downloadJSON and
+// inserts the pricing in the database.
+func importResult(reader io.ReadCloser, logger jsonlog.Logger, sqlTx models.XODB) error {
+	defer reader.Close()
+
+	decoder := json.NewDecoder(reader)
+	var bodyMap EC2Pricing
+	if err := decoder.Decode(&bodyMap); err != nil {
+		return err
+	}
+	for _, product := range bodyMap.Products {
+		if product.ProductFamily == "Compute Instance" {
+			var dbAwsProduct models.AwsProductEc2
+
+			dbAwsProduct.Sku = string(product.Sku)
+			if err := storeAttributes(&product.Attributes, &dbAwsProduct, logger, sqlTx); err != nil {
+				return err
+			}
+			if err := dbAwsProduct.Insert(sqlTx); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func saveLastFetch(lastFetch *models.AwsFetchPricing, newEtag string, sqlTx models.XODB) error {
+	if lastFetch != nil {
+		lastFetch.Delete(sqlTx)
+	}
+	dbAfp := models.AwsFetchPricing{
+		Product: "ec2",
+		Etag:    newEtag,
+	}
+	return dbAfp.Save(sqlTx)
+}
+
+// downloadJSON requests AWS' API and returns the body after
+// checking if there's already the same version in the database.
+func downloadJSON(logger jsonlog.Logger, sqlTx models.XODB) (io.ReadCloser, error) {
+	hc := http.Client{}
+	req, err := http.NewRequest("GET", config.UrlEC2Pricing, nil)
+	if err != nil {
+		return nil, err
+	}
+	lastFetch, err := models.AwsFetchPricingByProduct(db.Db, "ec2")
+	if err == nil {
+		req.Header.Add("If-None-Match", lastFetch.Etag)
+	}
+	res, err := hc.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	if res.StatusCode == 304 {
+		logger.Info("JSON Downloading halted: Already exists with ETag.\n", lastFetch.Etag)
+		return nil, nil
+	}
+	if err = saveLastFetch(lastFetch, res.Header["Etag"][0], sqlTx); err != nil {
+		logger.Error("Error when saving the dbAfp", err.Error())
+		return nil, err
+	}
+	return res.Body, nil
+}
+
+// ImportEC2Pricing downloads the EC2 Pricing from AWS and
+// store it in the database.
+func ImportEC2Pricing() error {
+	logger := jsonlog.DefaultLogger
+	sqlTx, err := db.Db.BeginTx(context.Background(), nil)
+	if err != nil {
+		logger.Error("Error when beginning sqlTx", err.Error())
+		return err
+	}
+	res, err := downloadJSON(logger, sqlTx)
```

Comment:
Here, if there are no errors, you
1. Open a connection to download a JSON document
2. Purge a table in the database
3. Consume the JSON document

You can’t know for sure that the database purge will end before the connection times out, so you have two possible solutions :
1. Insert products along with the etag (or some other id) for the document they came from, with an `INSERT … ON DUPLICATE KEY UPDATE`, then remove from the database all records with another etag (this is my preferred option).
2. Immediately consume the JSON document by writing it to a temporary file, purge the database, then consume the temporary file (don’t forget to remove it afterwards).